### PR TITLE
Storage Encryption Key Rotation Sample using Veneer + Tests

### PR DESCRIPTION
**Pull request by ryanmats**
Cloud Storage encryption key rotation sample in veneer:
- Implemented rotate_encryption_key
- Added test_rotate_encryption_key and made another minor fix to test file
- Verified that all tests pass both programmatically and from command line

Feature Request 32908100 in Buganizer

### CODE REVIEW

File: storage/cloud-client/encryption.py [hunk 1]
```diff
@@ -97,8 +97,13 @@ def rotate_encryption_key(bucket_name, blob_name, base64_encryption_key,
                           base64_new_encryption_key):
     """Performs a key rotation by re-writing an encrypted blob with a new
     encryption key."""
-    raise NotImplementedError(
-        'This is currently not available using the Cloud Client Library.')
+    storage_client = storage.Client()
+    bucket = storage_client.get_bucket(bucket_name)
+    blob = Blob(blob_name, bucket, encryption_key=base64.b64decode(base64_encryption_key))
```
Comment (jonparrott) [inline storage/cloud-client/encryption.py]: Maybe do `source_blob` and `destination_blob`. Add a comment explaining that they refer to the same storage object.

File: storage/cloud-client/encryption.py [hunk 2]
```diff
@@ -97,8 +97,13 @@ def rotate_encryption_key(bucket_name, blob_name, base64_encryption_key,
...
+    blob_new = Blob(blob_name, bucket, encryption_key=base64.b64decode(base64_new_encryption_key))
```
Comment (jonparrott) [inline storage/cloud-client/encryption.py]: Go ahead and do:

```
current_encryption_key = base64.b64decode(b64_encryption_key)
```

etc.

File: storage/cloud-client/encryption.py [hunk 3]
```diff
@@ -97,8 +97,13 @@ def rotate_encryption_key(bucket_name, blob_name, base64_encryption_key,
...
+    token, bytes_rewritten, total_bytes = blob_new.rewrite(blob)
+    while token is not None:
```
Comment (jonparrott) [inline storage/cloud-client/encryption.py]: Blank newline above `while`.

File: storage/cloud-client/encryption.py [hunk 4]
```diff
@@ -97,8 +97,13 @@ def rotate_encryption_key(bucket_name, blob_name, base64_encryption_key,
...
+        token, bytes_rewritten, total_bytes = blob_new.rewrite(blob, token)
```
Comment (jonparrott) [inline storage/cloud-client/encryption.py]: Print out something after this is complete?

File: storage/cloud-client/encryption_test.py [hunk 5]
```diff
@@ -16,13 +16,17 @@
 import tempfile
 
 from google.cloud import storage
+from google.cloud.storage import Blob
 import pytest
 
 import encryption
 
-TEST_ENCRYPTION_KEY = 'brtJUWneL92g5q0N2gyDSnlPSYAiIVZ/cWgjyZNeMy0='
+TEST_ENCRYPTION_KEY = '0YSyzDHlx5QAqSDOnrpqktm3+RkqiP1udwjrk3d7SGk='
```
Comment (jonparrott) [inline storage/cloud-client/encryption_test.py]: Why did this change?

File: storage/cloud-client/encryption_test.py [hunk 6]
```diff
@@ -65,3 +68,21 @@ def test_download_blob(test_blob, cloud_config):
...
+
+
+def test_rotate_encryption_key(test_blob, cloud_config):
+    test_blob_name, test_blob_content = test_blob
+    encryption.rotate_encryption_key(
+        cloud_config.storage_bucket,
+        test_blob_name,
+        TEST_ENCRYPTION_KEY,
+        TEST_ENCRYPTION_KEY_2)
+    with tempfile.NamedTemporaryFile() as dest_file:
```
Comment (jonparrott) [inline storage/cloud-client/encryption_test.py]: blank newline above `with`.

File: storage/cloud-client/encryption.py (hunk 1 above)
Comment (ryanmats) [inline storage/cloud-client/encryption.py]: done

File: storage/cloud-client/encryption.py (hunk 2 above)
Comment (ryanmats) [inline storage/cloud-client/encryption.py]: done

File: storage/cloud-client/encryption.py (hunk 3 above)
Comment (ryanmats) [inline storage/cloud-client/encryption.py]: done

File: storage/cloud-client/encryption.py (hunk 4 above)
Comment (ryanmats) [inline storage/cloud-client/encryption.py]: done

File: storage/cloud-client/encryption_test.py (hunk 6 above)
Comment (ryanmats) [inline storage/cloud-client/encryption_test.py]: done

File: storage/cloud-client/encryption_test.py (hunk 5 above)
Comment (ryanmats) [inline storage/cloud-client/encryption_test.py]: for some reason i had thought the old key wasn't working. seems to work now. changed it back!

File: storage/cloud-client/encryption.py [hunk 7]
```diff
@@ -97,8 +97,20 @@ def rotate_encryption_key(bucket_name, blob_name, base64_encryption_key,
                           base64_new_encryption_key):
     """Performs a key rotation by re-writing an encrypted blob with a new
     encryption key."""
-    raise NotImplementedError(
-        'This is currently not available using the Cloud Client Library.')
+    storage_client = storage.Client()
+    bucket = storage_client.get_bucket(bucket_name)
+    current_encryption_key = base64.b64decode(base64_encryption_key)
+    new_encryption_key = base64.b64decode(base64_new_encryption_key)
+    # Both source_blob and destination_blob refer to the same storage object,
```
Comment (jonparrott) [inline storage/cloud-client/encryption.py]: Blank newline above here?

File: storage/cloud-client/encryption.py [hunk 8]
```diff
@@ -97,8 +97,20 @@ def rotate_encryption_key(bucket_name, blob_name, base64_encryption_key,
...
+    # but destination_blob has the new encryption key.
+    source_blob = Blob(blob_name, bucket, encryption_key=current_encryption_key)
+    destination_blob = Blob(blob_name, bucket, encryption_key=new_encryption_key)
+    token, bytes_rewritten, total_bytes = destination_blob.rewrite(source_blob)
```
Comment (jonparrott) [inline storage/cloud-client/encryption.py]: Can you do:

```python
token = None

while True:
    token, bytes_rewritten, total_bytes = destination_blob.rewrite(source_blob, token=token)
    if token is None:
        break
```

File: storage/cloud-client/encryption.py (hunk 7 above)
Comment (ryanmats) [inline storage/cloud-client/encryption.py]: done

**Comment by jonparrott** [comment]
@dpebot can you merge when travis passes?

**Pull request merged**